// ccolor/app.js
// Import.
const colors = require('./colors');

/**
 * Console color.
 */
class ConsoleColor {
  /**
   * Colors.
   */
  static get Colors() { return colors; }

  /**
   * Set color.
   * @param {string} text Text
   * @param {...string} colorsToUse Colors to use.
   * @returns {string} Colored text.
   */
  static setColor(text = '', ...colorsToUse) {
    return `${colorsToUse.join('')}${text}${colors.reset}`;
  }

  /**
   * Black.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static black(text = '') {
    return ConsoleColor.setColor(text, colors.foreground.black);
  }

  /**
   * White.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static white(text = '') {
    return ConsoleColor.setColor(text, colors.foreground.white);
  }

  /**
   * Blue.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static blue(text = '') {
    return ConsoleColor.setColor(text, colors.foreground.blue);
  }

  /**
   * Green.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static green(text = '') {
    return ConsoleColor.setColor(text, colors.foreground.green);
  }

  /**
   * Yellow.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static yellow(text = '') {
    return ConsoleColor.setColor(text, colors.foreground.yellow);
  }

  /**
   * Red.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static red(text = '') {
    return ConsoleColor.setColor(text, colors.foreground.red);
  }

  /**
   * Magenta.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static magenta(text = '') {
    return ConsoleColor.setColor(text, colors.foreground.magenta);
  }

  /**
   * Cyan.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static cyan(text = '') {
    return ConsoleColor.setColor(text, colors.foreground.cyan);
  }

  /**
   * Black background.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static blackBackground(text = '') {
    return ConsoleColor.setColor(text, colors.background.black);
  }

  /**
   * White background.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static whiteBackground(text = '') {
    return ConsoleColor.setColor(text, colors.background.white);
  }

  /**
   * Blue background.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static blueBackground(text = '') {
    return ConsoleColor.setColor(text, colors.background.blue);
  }

  /**
   * Green background.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static greenBackground(text = '') {
    return ConsoleColor.setColor(text, colors.background.green);
  }

  /**
   * Yellow background.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static yellowBackground(text = '') {
    return ConsoleColor.setColor(text, colors.background.yellow);
  }

  /**
   * Red background.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static redBackground(text = '') {
    return ConsoleColor.setColor(text, colors.background.red);
  }

  /**
   * Magenta background.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static magentaBackground(text = '') {
    return ConsoleColor.setColor(text, colors.background.magenta);
  }

  /**
   * Cyan background.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static cyanBackground(text = '') {
    return ConsoleColor.setColor(text, colors.background.cyan);
  }

  /**
   * Bright style.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static brightStyle(text = '') {
    return ConsoleColor.setColor(text, colors.style.bright);
  }

  /**
   * Dim style.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static dimStyle(text = '') {
    return ConsoleColor.setColor(text, colors.style.dim);
  }

  /**
   * Underscore style.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static underscoreStyle(text = '') {
    return ConsoleColor.setColor(text, colors.style.underscore);
  }

  /**
   * Blink style.
   * @param {string} text Text.
   * @returns {string} Colored text.
   */
  static blinkStyle(text = '') {
    return ConsoleColor.setColor(text, colors.style.blink);
  }
}

// Export.
module.exports = ConsoleColor;
